pkg/protocol/handshake/message_server_key_exchange.go

```diff
@@ -27,19 +31,22 @@ func (m MessageServerKeyExchange) Type() Type {
 
 // Marshal encodes the Handshake
 func (m *MessageServerKeyExchange) Marshal() ([]byte, error) {
+	var out []byte
 	if m.IdentityHint != nil {
-		out := append([]byte{0x00, 0x00}, m.IdentityHint...)
+		out = append([]byte{0x00, 0x00}, m.IdentityHint...)
 		binary.BigEndian.PutUint16(out, uint16(len(out)-2))
-		return out, nil
 	}
 
-	out := []byte{byte(m.EllipticCurveType), 0x00, 0x00}
-	binary.BigEndian.PutUint16(out[1:], uint16(m.NamedCurve))
+	if m.EllipticCurveType == 0 || len(m.PublicKey) == 0 {
+		return out, nil
+	}
+	out = append(out, byte(m.EllipticCurveType), 0x00, 0x00)
+	binary.BigEndian.PutUint16(out[len(out)-2:], uint16(m.NamedCurve))
 
 	out = append(out, byte(len(m.PublicKey)))
 	out = append(out, m.PublicKey...)
 
-	if m.HashAlgorithm == hash.None && m.SignatureAlgorithm == signature.Anonymous && len(m.Signature) == 0 {
+	if m.HashAlgorithm == hash.None || m.SignatureAlgorithm == signature.Anonymous || len(m.Signature) == 0 {
```

Comment:
Are these OR cases all valid?

Reply:
@jkralik

Reply:
Yes. Each field must be filled in otherwise the signature is not valid. (Used in the ANON cipher - all mentioned members are set as is in the condition)

Reply:
Isn't it better to return error on invalid cases?

Reply:
So I added checks for invalid cases.